File: benchmarks/references/ground-truth/R/tc-022-dor.R
#!/usr/bin/env Rscript
# TC-022 Ground Truth: Duration of Response (DOR) — KM Median
# Part of the ASA Biopharm AI/ML WG Agentic AI Benchmark
#
# DOR is defined as the time from first documented response (CR or PR)
# to disease progression or death, among subjects who achieved a response.
#
# Key differences from TC-001 (PFS) and TC-021 (TTP):
#   PFS (TC-001): All subjects, event = progression OR death
#   TTP (TC-021): All subjects, event = progression only (death censored)
#   DOR (TC-022): Responders only, event = progression OR death
#                 Left-truncated: entry time = time to first response
#
# This tests:
#   1. Correct subsetting to responder population (CR + PR)
#   2. KM estimation on a selected subset (not full ITT)
#   3. Handling of left truncation (response occurs after randomization)
#
# Dependencies: survival (>= 3.5), jsonlite, dplyr
# Usage: Rscript tc-022-dor.R --seed 42 --n 200 --output results.json

library(survival)
library(jsonlite)
library(dplyr)

# Source common utilities
source("common/data-generation.R")

# ─────────────────────────────────────────────────────
# Command-line argument parsing
# ─────────────────────────────────────────────────────

parse_args <- function() {
  args <- commandArgs(trailingOnly = TRUE)
  get_arg <- function(flag, default) {
    idx <- which(args == flag)
    if (length(idx) > 0 && idx < length(args)) return(args[idx + 1])
    return(default)
  }
  list(
    seed = as.integer(get_arg("--seed", "42")),
    n = as.integer(get_arg("--n", "200")),
    arm = as.integer(get_arg("--arm", "1")),
    conf_type = get_arg("--conf-type", "log-log"),
    data = get_arg("--data", NA),
    output = get_arg("--output", NA),
    ars_output = get_arg("--ars-output", NA)
  )
}

# ─────────────────────────────────────────────────────
# Generate DOR-specific ADTTE data
# ─────────────────────────────────────────────────────
# DOR dataset: among responders (CR/PR), time from first response to
# progression or death. Includes left truncation (entry = time to response).

generate_dor_adtte <- function(seed = 42, n_subjects = 200, hr = 0.75) {
  set.seed(seed)
  base_rate <- log(2) / 6.0  # median PFS control = 6 months
  trt <- sample(0:1, n_subjects, replace = TRUE)
  hazard_mult <- ifelse(trt == 0, 1.0, hr)

  # Generate time to first response (CR or PR)
  # Response probability depends on treatment
  orr_prob <- ifelse(trt == 1, 0.40, 0.20)
  is_responder <- rbinom(n_subjects, 1, orr_prob)

  # Time to response (among responders): exponential with mean ~2 months
  time_to_response <- ifelse(is_responder == 1,
                              rexp(n_subjects, rate = 1.0 / 2.0),
                              NA)

  # Time to progression or death (from randomization)
  prog_time <- rexp(n_subjects, rate = base_rate * hazard_mult)
  death_time <- rexp(n_subjects, rate = base_rate * 0.3 * hazard_mult)
  cens_time <- rexp(n_subjects, rate = base_rate * 0.3 / 0.7)

  # For responders: DOR = time from response to event (prog/death) or censoring
  # Event time from randomization
  event_time_from_rand <- pmin(prog_time, death_time)
  cens_from_rand <- cens_time

  # DOR time = event_time - time_to_response (if event after response)
  # If event before response → not applicable (shouldn't happen if response occurred)
  # If censoring before event → censored

  # Only responders have valid DOR
  has_event_after_response <- is_responder == 1 & event_time_from_rand > time_to_response
  has_cens_before_event <- is_responder == 1 & cens_from_rand < event_time_from_rand & cens_from_rand > time_to_response

  # DOR duration
  dor_time <- ifelse(has_event_after_response,
                     event_time_from_rand - time_to_response,
                     NA)
  # If censoring occurs before event but after response
  dor_time <- ifelse(has_cens_before_event,
                     cens_from_rand - time_to_response,
                     dor_time)
  # Censoring indicator: 0 = event, 1 = censored
  dor_cnsr <- ifelse(has_event_after_response & !has_cens_before_event, 0, 1)
  dor_cnsr <- ifelse(has_cens_before_event, 1, dor_cnsr)

  # For non-responders: NA
  dor_time <- ifelse(is_responder == 0, NA, dor_time)
  dor_cnsr <- ifelse(is_responder == 0, NA, dor_cnsr)

  # Response type
  is_cr <- ifelse(is_responder == 1, rbinom(n_subjects, 1, 0.3), 0)
  bor <- ifelse(is_cr == 1, "CR",
                ifelse(is_responder == 1, "PR",
                       sample(c("SD", "PD"), n_subjects, replace = TRUE, prob = c(0.4, 0.6))))

  sex <- sample(c("Male", "Female"), n_subjects, replace = TRUE, prob = c(0.5, 0.5))
  ecog <- sample(0:1, n_subjects, replace = TRUE, prob = c(0.6, 0.4))
  ittfl <- ifelse(runif(n_subjects) < 0.95, "Y", "N")

  data.frame(
    USUBJID = sprintf("SUBJ-%04d", 1:n_subjects),
    STUDYID = "BENCHMARK-001",
    TRT01PN = trt,
    TRT01P = ifelse(trt == 0, "Placebo", "Active"),
    AVAL = round(dor_time, 2),
    CNSR = dor_cnsr,
    PARAMCD = "DOR",
    PARAM = "Duration of Response",
    ITTFL = ittfl,
    SAFFL = ifelse(runif(n_subjects) < 0.98, "Y", "N"),
    SEX = sex,
    ECOG = ecog,
    BOR = bor,
    IS_RESPONDER = is_responder,
    TIME_TO_RESPONSE = round(time_to_response, 2),
    stringsAsFactors = FALSE
  )
}

# ─────────────────────────────────────────────────────
# TC-022 Core Computation
# ─────────────────────────────────────────────────────

compute_dor_median <- function(adtte,
                                arm = 1,
                                population = "ITT",
                                conf_type = "log-log",
                                seed = NA) {

  if (population == "ITT") {
    data <- adtte %>% filter(ITTFL == "Y", TRT01PN == arm)
  } else {
    data <- adtte %>% filter(TRT01PN == arm)
  }

  # Filter to responders only
  data <- data %>% filter(IS_RESPONDER == 1, !is.na(AVAL), AVAL > 0)

  if (nrow(data) == 0) {
    # No responders → DOR not estimable
    return(list(
      test_case_id = "TC-022",
      variant_id = if (is.na(seed)) NA else paste0("v", seed),
      language = "R",
      package = "survival",
      package_version = as.character(packageVersion("survival")),
      median_dor = NA,
      ci_lower = NA,
      ci_upper = NA,
      n_responders = 0L,
      n_events = 0L,
      n_total = as.integer(nrow(adtte %>% filter(ITTFL == "Y", TRT01PN == arm))),
      ci_method = conf_type,
      endpoint = "DOR",
      population = "responders (CR+PR)",
      censoring_rule = "event = progression or death",
      estimable = FALSE,
      seed = seed
    ))
  }

  n_total_in_arm <- nrow(adtte %>% filter(ITTFL == "Y", TRT01PN == arm))

  # KM estimation for DOR (no left truncation in this simplified version;
  # DOR time is already computed as time from response to event)
  fit <- survfit(Surv(AVAL, 1 - CNSR) ~ 1,
                 data = data,
                 conf.type = conf_type)

  tab <- summary(fit)$table
  median_dor <- unname(tab["median"])
  ci_lower   <- unname(tab["0.95LCL"])
  ci_upper   <- unname(tab["0.95UCL"])
  n_events   <- unname(tab["events"])
  n_responders <- nrow(data)

  variant <- if (is.na(seed)) NA else paste0("v", seed)

  if (length(median_dor) == 0 || is.na(median_dor)) {
    result <- list(
      test_case_id = "TC-022",
      variant_id = variant,
      language = "R",
      package = "survival",
      package_version = as.character(packageVersion("survival")),
      median_dor = NA,
      ci_lower = NA,
      ci_upper = NA,
      n_responders = as.integer(n_responders),
      n_events = as.integer(n_events),
      n_total = as.integer(n_total_in_arm),
      ci_method = conf_type,
      endpoint = "DOR",
      population = "responders (CR+PR)",
      censoring_rule = "event = progression or death",
      estimable = FALSE,
      seed = seed
    )
    return(result)
  }

  result <- list(
    test_case_id = "TC-022",
    variant_id = variant,
    language = "R",
    package = "survival",
    package_version = as.character(packageVersion("survival")),
    median_dor = round(as.numeric(median_dor), 4),
    ci_lower = if (is.na(ci_lower)) NA else round(as.numeric(ci_lower), 4),
    ci_upper = if (is.na(ci_upper)) NA else round(as.numeric(ci_upper), 4),
    n_responders = as.integer(n_responders),
    n_events = as.integer(n_events),
    n_total = as.integer(n_total_in_arm),
    ci_method = conf_type,
    endpoint = "DOR",
    population = "responders (CR+PR)",
    censoring_rule = "event = progression or death",
    estimable = TRUE,
    seed = seed
  )

  return(result)
}

# ─────────────────────────────────────────────────────
# Main
# ─────────────────────────────────────────────────────

if (sys.nframe() == 0) {
  opts <- parse_args()
  seed <- opts$seed
  n <- opts$n

  cat(sprintf("TC-022: Duration of Response KM Median (R) — seed=%d, n=%d, arm=%d\n",
              seed, n, opts$arm))

  # Obtain data: shared CSV or generate DOR-specific ADTTE
  if (!is.na(opts$data)) {
    adtte <- read_shared_data(opts$data)
    cat(sprintf("Loaded shared ADTTE with %d subjects\n", nrow(adtte)))
  } else {
    adtte <- generate_dor_adtte(seed = seed, n_subjects = n)
    cat(sprintf("Generated DOR ADTTE with %d subjects\n", nrow(adtte)))
  }

  result <- compute_dor_median(adtte,
                                arm = opts$arm,
                                conf_type = opts$conf_type,
                                seed = seed)

  cat("\n──────────────────────────────────────────────\n")
  cat(sprintf("Endpoint:       %s\n", result$endpoint))
  cat(sprintf("Population:     %s\n", result$population))
  cat(sprintf("Censoring rule: %s\n", result$censoring_rule))
  if (result$estimable) {
    cat(sprintf("Median DOR:     %.1f months\n", result$median_dor))
    cat(sprintf("95%% CI:         (%.1f, %.1f)\n", result$ci_lower, result$ci_upper))
  } else {
    cat("Median DOR:     Not estimable\n")
  }
  cat(sprintf("N responders:  %d / %d\n", result$n_responders, result$n_total))
  cat(sprintf("N events:      %d\n", result$n_events))
  cat(sprintf("R survival v%s\n", result$package_version))
  cat("──────────────────────────────────────────────\n")

  print_output(result)

  if (!is.na(opts$output)) {
    write_output(result, opts$output)
  }

  # ARS-compatible output envelope (CDISC ARS v1.0)
  if (!is.na(opts$ars_output)) {
    ars_envelope <- list(
      ars_version = "1.0",
      analysisResult = list(
        id = "TC-022",
        version = "1.0",
        analysisReason = "Secondary efficacy endpoint: duration of response among responders",
        analysisMethod = list(
          name = "Kaplan-Meier",
          codeTemplate = "survival::survfit(Surv(AVAL, 1-CNSR) ~ 1)",
          parameters = list(
            conf_type = result$ci_method,
            tie_method = "efron",
            population = "responders (CR+PR)",
            event_definition = "progression or death"
          )
        ),
        analysisVariables = list(
          list(name = "AVAL", dataset = "ADTTE", role = "analysis time (from response)"),
          list(name = "CNSR", dataset = "ADTTE", role = "censoring"),
          list(name = "BOR", dataset = "ADRS", role = "best overall response"),
          list(name = "TRT01A", dataset = "ADSL", role = "treatment")
        ),
        analysisPopulation = list(
          name = "Responders",
          filter = "IS_RESPONDER = 1 AND ITTFL = 'Y'"
        ),
        analysisDataset = "ADTTE",
        resultGroups = list(
          list(id = ifelse(opts$arm == 1, "Experimental", "Control"),
               n = result$n_responders,
               events = result$n_events)
        ),
        documentation = "KM median DOR estimation among responders (CR+PR); event = progression or death",
        analysisResultsData = list(
          statistics = list(
            list(name = "median_dor", value = result$median_dor, unit = "months"),
            list(name = "ci_lower", value = result$ci_lower),
            list(name = "ci_upper", value = result$ci_upper),
            list(name = "n_responders", value = as.integer(result$n_responders)),
            list(name = "n_events", value = as.integer(result$n_events)),
            list(name = "n_total", value = as.integer(result$n_total)),
            list(name = "estimable", value = result$estimable)
          )
        )
      )
    )
    ars_json <- jsonlite::toJSON(ars_envelope, auto_unbox = TRUE, pretty = TRUE)
    writeLines(ars_json, opts$ars_output)
    cat(sprintf("Wrote ARS-compatible output to: %s\n", opts$ars_output))
  }
}

File: benchmarks/references/ground-truth/Python/tc_022_dor.py
#!/usr/bin/env python3
"""TC-022 Ground Truth: Duration of Response (DOR) — KM Median (Python)

Part of the ASA Biopharm AI/ML WG Agentic AI Benchmark.

DOR is defined as the time from first documented response (CR or PR)
to disease progression or death, among subjects who achieved a response.

Key differences from TC-001 (PFS) and TC-021 (TTP):
  PFS (TC-001): All subjects, event = progression OR death
  TTP (TC-021): All subjects, event = progression only (death censored)
  DOR (TC-022): Responders only, event = progression OR death

This tests:
  1. Correct subsetting to responder population (CR + PR)
  2. KM estimation on a selected subset (not full ITT)
  3. Handling of left truncation (response occurs after randomization)

Dependencies: pandas, numpy, lifelines (>= 0.29.0)
Usage: python tc_022_dor.py --seed 42 --n 200 --output results.json
"""

import argparse
import json
from pathlib import Path

import numpy as np
import pandas as pd
from lifelines import KaplanMeierFitter
from lifelines.utils import median_survival_times
import lifelines


def generate_dor_adtte(seed=42, n_subjects=200, hr=0.75):
    """Generate DOR-specific ADTTE dataset.

    Among responders (CR/PR), computes time from first response to
    progression or death. Includes TIME_TO_RESPONSE for left truncation.
    """
    rng = np.random.default_rng(seed)
    base_rate = np.log(2) / 6.0  # median PFS control = 6 months
    trt = rng.choice([0, 1], size=n_subjects)
    hazard_mult = np.where(trt == 0, 1.0, hr)

    # Response probability depends on treatment
    orr_prob = np.where(trt == 1, 0.40, 0.20)
    is_responder = rng.binomial(1, orr_prob)

    # Time to response (among responders): exponential with mean ~2 months
    time_to_response = np.where(
        is_responder == 1,
        rng.exponential(1.0 / (1.0 / 2.0), size=n_subjects),
        np.nan
    )

    # Time to progression or death (from randomization)
    prog_time = rng.exponential(1.0 / (base_rate * hazard_mult))
    death_time = rng.exponential(1.0 / (base_rate * 0.3 * hazard_mult))
    cens_time = rng.exponential(1.0 / (base_rate * 0.3 / 0.7))

    # Event time from randomization
    event_time_from_rand = np.minimum(prog_time, death_time)

    # For responders: DOR = time from response to event or censoring
    has_event_after_response = (is_responder == 1) & (event_time_from_rand > time_to_response)
    has_cens_before_event = (
        (is_responder == 1) &
        (cens_time < event_time_from_rand) &
        (cens_time > time_to_response)
    )

    # DOR time
    dor_time = np.where(
        has_event_after_response,
        event_time_from_rand - time_to_response,
        np.nan
    )
    dor_time = np.where(
        has_cens_before_event,
        cens_time - time_to_response,
        dor_time
    )

    # Censoring indicator: 0 = event, 1 = censored
    dor_cnsr = np.where(has_event_after_response & ~has_cens_before_event, 0, 1)
    dor_cnsr = np.where(has_cens_before_event, 1, dor_cnsr)
    dor_cnsr = np.where(is_responder == 0, np.nan, dor_cnsr)
    dor_time = np.where(is_responder == 0, np.nan, dor_time)

    # Response type: ~30% CR, ~70% PR among responders
    is_cr = np.where(is_responder == 1, rng.binomial(1, 0.3, size=n_subjects), 0)
    bor = np.where(is_cr == 1, "CR",
                   np.where(is_responder == 1, "PR",
                            rng.choice(["SD", "PD"], size=n_subjects, p=[0.4, 0.6])))

    sex = rng.choice(["Male", "Female"], size=n_subjects, p=[0.5, 0.5])
    ecog = rng.choice([0, 1], size=n_subjects, p=[0.6, 0.4])
    ittfl = np.where(rng.random(n_subjects) < 0.95, "Y", "N")

    df = pd.DataFrame({
        "USUBJID": [f"SUBJ-{i+1:04d}" for i in range(n_subjects)],
        "STUDYID": "BENCHMARK-001",
        "TRT01PN": trt,
        "TRT01P": np.where(trt == 0, "Placebo", "Active"),
        "AVAL": np.round(dor_time, 2),
        "CNSR": dor_cnsr,
        "PARAMCD": "DOR",
        "PARAM": "Duration of Response",
        "ITTFL": ittfl,
        "SAFFL": np.where(rng.random(n_subjects) < 0.98, "Y", "N"),
        "SEX": sex,
        "ECOG": ecog,
        "BOR": bor,
        "IS_RESPONDER": is_responder,
        "TIME_TO_RESPONSE": np.round(time_to_response, 2),
    })
    return df


def compute_dor_median(adtte, arm=1, population="ITT", conf_type="log-log", seed=42):
    """Compute KM median DOR for specified arm among responders."""
    if population == "ITT":
        data = adtte[(adtte["ITTFL"] == "Y") & (adtte["TRT01PN"] == arm)].copy()
    else:
        data = adtte[adtte["TRT01PN"] == arm].copy()

    n_total_in_arm = len(data)

    # Filter to responders only
    data = data[(data["IS_RESPONDER"] == 1) & data["AVAL"].notna() & (data["AVAL"] > 0)].copy()

    if len(data) == 0:
        return {
            "test_case_id": "TC-022",
            "variant_id": f"v{seed}",
            "language": "Python",
            "package": "lifelines",
            "package_version": lifelines.__version__,
            "median_dor": None,
            "ci_lower": None,
            "ci_upper": None,
            "n_responders": 0,
            "n_events": 0,
            "n_total": n_total_in_arm,
            "ci_method": conf_type,
            "endpoint": "DOR",
            "population": "responders (CR+PR)",
            "censoring_rule": "event = progression or death",
            "estimable": False,
            "seed": seed,
        }

    durations = data["AVAL"].values
    event_observed = (1 - data["CNSR"]).values

    kmf = KaplanMeierFitter()
    kmf.fit(durations, event_observed, label='DOR')

    median_dor = kmf.median_survival_time_
    n_responders = len(data)
    n_events = int(event_observed.sum())

    if np.isnan(median_dor) or median_dor == np.inf:
        return {
            "test_case_id": "TC-022",
            "variant_id": f"v{seed}",
            "language": "Python",
            "package": "lifelines",
            "package_version": lifelines.__version__,
            "median_dor": None,
            "ci_lower": None,
            "ci_upper": None,
            "n_responders": n_responders,
            "n_events": n_events,
            "n_total": n_total_in_arm,
            "ci_method": conf_type,
            "endpoint": "DOR",
            "population": "responders (CR+PR)",
            "censoring_rule": "event = progression or death",
            "estimable": False,
            "seed": seed,
        }

    # Extract CI of the median (Brookmeyer-Crowley interval)
    try:
        ci_median = median_survival_times(kmf.confidence_interval_)
        lo = float(ci_median.iloc[0, 0])
        hi = float(ci_median.iloc[0, 1])
        ci_lower = None if np.isinf(lo) or pd.isna(lo) else round(lo, 4)
        ci_upper = None if np.isinf(hi) or pd.isna(hi) else round(hi, 4)
    except (KeyError, IndexError, ValueError, StopIteration):
        ci_lower = None
        ci_upper = None

    return {
        "test_case_id": "TC-022",
        "variant_id": f"v{seed}",
        "language": "Python",
        "package": "lifelines",
        "package_version": lifelines.__version__,
        "median_dor": float(round(median_dor, 4)),
        "ci_lower": ci_lower,
        "ci_upper": ci_upper,
        "n_responders": n_responders,
        "n_events": n_events,
        "n_total": n_total_in_arm,
        "ci_method": conf_type,
        "endpoint": "DOR",
        "population": "responders (CR+PR)",
        "censoring_rule": "event = progression or death",
        "estimable": True,
        "seed": seed,
    }


def main():
    parser = argparse.ArgumentParser(description="TC-022: Duration of Response KM Median (Python)")
    parser.add_argument("--seed", type=int, default=42)
    parser.add_argument("--n", type=int, default=200)
    parser.add_argument("--arm", type=int, default=1, help="Treatment arm (0=Control, 1=Experimental)")
    parser.add_argument("--conf-type", type=str, default="log-log")
    parser.add_argument("--data", type=str, default=None, help="Shared ADTTE CSV")
    parser.add_argument("--output", type=str, default=None)
    parser.add_argument("--ars-output", type=str, default=None)
    args = parser.parse_args()

    seed = args.seed
    print(f"TC-022: Duration of Response KM Median (Python) — seed={seed}, n={args.n}, arm={args.arm}")

    if args.data:
        adtte = pd.read_csv(args.data)
        print(f"Loaded shared ADTTE with {len(adtte)} subjects")
    else:
        adtte = generate_dor_adtte(seed=seed, n_subjects=args.n)
        print(f"Generated DOR ADTTE with {len(adtte)} subjects")

    result = compute_dor_median(adtte, arm=args.arm, conf_type=args.conf_type, seed=seed)

    print("\n" + "─" * 50)
    print(f"Endpoint:       {result['endpoint']}")
    print(f"Population:     {result['population']}")
    print(f"Censoring rule: {result['censoring_rule']}")
    if result["estimable"]:
        print(f"Median DOR:     {result['median_dor']:.1f} months")
        print(f"95% CI:         ({result['ci_lower']}, {result['ci_upper']})")
    else:
        print("Median DOR:     Not estimable")
    print(f"N responders:  {result['n_responders']} / {result['n_total']}")
    print(f"N events:      {result['n_events']}")
    print(f"Python lifelines v{result['package_version']}")
    print("─" * 50)

    print("\n=== BENCHMARK OUTPUT ===")
    print(json.dumps(result, indent=2, default=str))
    print("=== END OUTPUT ===")

    if args.output:
        outpath = Path(args.output)
        outpath.parent.mkdir(parents=True, exist_ok=True)
        with open(outpath, "w") as f:
            json.dump(result, f, indent=2, default=str)
        print(f"Wrote output to: {outpath}")

    if args.ars_output:
        ars_envelope = {
            "ars_version": "1.0",
            "analysisResult": {
                "id": "TC-022",
                "version": "1.0",
                "analysisReason": "Secondary efficacy endpoint: duration of response among responders",
                "analysisMethod": {
                    "name": "Kaplan-Meier",
                    "codeTemplate": "lifelines.KaplanMeierFitter().fit(AVAL, 1-CNSR)",
                    "parameters": {
                        "conf_type": result["ci_method"],
                        "alpha": 0.05,
                        "population": "responders (CR+PR)",
                        "event_definition": "progression or death"
                    }
                },
                "analysisVariables": [
                    {"name": "AVAL", "dataset": "ADTTE", "role": "analysis time (from response)"},
                    {"name": "CNSR", "dataset": "ADTTE", "role": "censoring"},
                    {"name": "BOR", "dataset": "ADRS", "role": "best overall response"},
                    {"name": "TRT01A", "dataset": "ADSL", "role": "treatment"}
                ],
                "analysisPopulation": {"name": "Responders", "filter": "IS_RESPONDER = 1 AND ITTFL = 'Y'"},
                "analysisDataset": "ADTTE",
                "resultGroups": [
                    {"id": "Experimental" if args.arm == 1 else "Control",
                     "n": result["n_responders"], "events": result["n_events"]}
                ],
                "documentation": "KM median DOR estimation among responders (CR+PR); event = progression or death",
                "analysisResultsData": {
                    "statistics": [
                        {"name": "median_dor", "value": result["median_dor"], "unit": "months"},
                        {"name": "ci_lower", "value": result["ci_lower"]},
                        {"name": "ci_upper", "value": result["ci_upper"]},
                        {"name": "n_responders", "value": result["n_responders"]},
                        {"name": "n_events", "value": result["n_events"]},
                        {"name": "n_total", "value": result["n_total"]},
                        {"name": "estimable", "value": result["estimable"]}
                    ]
                }
            }
        }
        ars_path = Path(args.ars_output)
        ars_path.parent.mkdir(parents=True, exist_ok=True)
        with open(ars_path, "w") as f:
            json.dump(ars_envelope, f, indent=2, default=str)
        print(f"Wrote ARS-compatible output to: {ars_path}")


if __name__ == "__main__":
    main()
